module/zfs/zil.c

```diff
@@ -1190,12 +1188,44 @@ zil_lwb_flush_vdevs_done(zio_t *zio)
 
 	mutex_exit(&zilog->zl_lock);
 
-	/*
-	 * Now that we've written this log block, we have a stable pointer
-	 * to the next block in the chain, so it's OK to let the txg in
-	 * which we allocated the next block sync.
-	 */
-	dmu_tx_commit(tx);
```

Comment:
@ahrens can you remind me why we chose to hold the TX for the entirety of the LWB write/flush? the comment mentions the next block in the chain.. why can't we let the TXG sync, until this block is flushed? may the TXG sync impact the next block in the chain, (potentially) causing some sort of corruption of the on-disk linked list?

I believe this PR is attempting to commit this TX at the time that we issue the LWB, rather than at the time the LWB write is flushed (current behavior), so I'm trying to think about why (or not) it'd be safe to commit the TX at the time we issue the write.. and trying to remember why we chose (need?) to wait to commit it unitl after the LWB has been flushed.

Reply:
I believe the goal was to ensure that zil_lwb_flush_vdevs_done() is called
before zil_sync(). zil_sync() is called in syncing context. Hold the txg open until zil_lwb_flush_vdevs_done() is called ensure the correct order().
In zil_lwb_flush_vdevs_done(), we set lwb->lwb_buf = NULL, so that zil_sync() can pick this lwb up and remove it. The order of the call is important.

Reply:
> I believe the goal was to ensure that zil_lwb_flush_vdevs_done() is called before zil_sync().

> In zil_lwb_flush_vdevs_done(), we set lwb->lwb_buf = NULL, so that zil_sync() can pick this lwb up and remove it.

But if `lwb_buf` is not NULL, then `zil_sync` will just skip that LWB (and all later LWBs in the `zl_lwb_list`).. for now.. in the next sync, it'll try to process `zl_lwb_list` again, and likely find that it can now be freed, right? meaning it's not critical to prevent `zil_sync` from being called until after the LWBs are flushed (so there must be another reason for doing this)...?

It's been a little while since I was in this code. so I'll have to refresh my memory..

Reply:
> But if `lwb_buf` is not NULL, then `zil_sync` will just skip that LWB (and all later LWBs in the `zl_lwb_list`).. for now.. in the next sync, it'll try to process `zl_lwb_list` again, and likely find that it can now be freed, right? meaning it's not critical to prevent `zil_sync` from being called until after the LWBs are flushed (so there must be another reason for doing this)...?
> 
> It's been a little while since I was in this code. so I'll have to refresh my memory..

I haven't looked in detail of this code path you mentioned above. 
However, I tried. I let zil_sync() without waiting zil_lwb_flush_vdevs_done(). zil_close() panics at ASSERT3P(lwb, ==, list_tail(&zilog->zl_lwb_list)). 
zil_close() expects all inflight lwbs should be settled after zl_dirty_max_txg or lwb_max_txg is synced.

Another potential issue, in zil_lwb_write_issue(), we call dsl_dataset_dirty(dmu_objset_ds(zilog->zl_os), tx). I believe zil_sync() only is called for dirty dataset in curretn txg. If we let lwb fly longer, it may not be picked up in next zil_sync().